src/Conf.ml

```diff
@@ -1009,25 +1054,41 @@ let parse_line config ~verbose ~from s =
   | _ -> Error (`Malformed s)
 
 let is_project_root dir =
-  List.exists project_root_witness ~f:(fun name ->
-      Caml.Sys.file_exists (Filename.concat dir name) )
-
-let rec collect_files ~dir acc =
-  let acc =
-    let filename = Filename.concat dir ".ocamlformat" in
-    if Caml.Sys.file_exists filename then filename :: acc else acc
-  in
-  if is_project_root dir && !disable_outside_project then (acc, Some dir)
-  else
-    let dir' = Filename.dirname dir in
-    if (not (String.equal dir dir')) && Caml.Sys.file_exists dir then
-      collect_files ~dir:dir' acc
-    else if !disable_outside_project then ([], None)
-    else (acc, None)
+  match root with
+  | Some root -> Fpath.equal dir root
+  | None ->
+      List.exists project_root_witness ~f:(fun name ->
+          match Bos.OS.Dir.exists Fpath.(append dir (v name)) with
+          | Ok r -> r
+          | Error _ -> false )
+
+let rec collect_files ~segs acc =
+  match segs with
+  | [] | [""] -> (acc, None)
+  | "" :: upper_segs -> collect_files ~segs:upper_segs acc
+  | _ :: upper_segs ->
+      let dir =
+        String.concat ~sep:Fpath.dir_sep (List.rev segs) |> Fpath.v
+      in
+      let acc =
+        let filename = Fpath.(append dir (v ".ocamlformat")) in
+        match Bos.OS.File.exists filename with
+        | Ok true -> filename :: acc
+        | _ -> acc
+      in
+      if is_project_root dir && disable_outside_detected_project then
+        (acc, Some dir)
+      else
+        let dir_exists =
+          match Bos.OS.Dir.exists dir with Ok r -> r | Error _ -> false
+        in
+        if dir_exists then collect_files ~segs:upper_segs acc
```

Comment:
I don't understand the situation where`dir_exists` is false, can you explain when that might happen and what the desired behavior is?

Reply:
This case should not exist, I simplified the function.